gpjax/decision_making/decision_maker.py

```diff
@@ -0,0 +1,254 @@
+# Copyright 2023 The GPJax Contributors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+from abc import (
+    ABC,
+    abstractmethod,
+)
+from dataclasses import (
+    dataclass,
+    field,
+)
+
+from beartype.typing import (
+    Callable,
+    Dict,
+    List,
+    Mapping,
+)
+import jax.random as jr
+
+from gpjax.dataset import Dataset
+from gpjax.decision_making.acquisition_functions import (
+    AbstractAcquisitionFunctionBuilder,
+)
+from gpjax.decision_making.acquisition_maximizer import AbstractAcquisitionMaximizer
+from gpjax.decision_making.posterior_handler import PosteriorHandler
+from gpjax.decision_making.search_space import AbstractSearchSpace
+from gpjax.decision_making.utils import FunctionEvaluator
+from gpjax.gps import AbstractPosterior
+from gpjax.typing import (
+    Array,
+    Float,
+    KeyArray,
+)
+
+
+@dataclass
+class AbstractDecisionMaker(ABC):
+    """
+    AbstractDecisionMaker abstract base class which handles the core decision loop. The
```

Comment:
maybe say what "core decision loop" means. I think its important to stress the sequential aspect